File: Dynamic_Programming/198_House_Robber/solution198.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int rob1 = 0, rob2 = 0;

        for (int n : nums) {
            int temp = max(n + rob1, rob2);
            rob1 = rob2;
            rob2 = temp;
        }
        
        return rob2;
    }
};

// Time Complexity:
// Space Complexity:

// Runtime1: 0ms, beats 100%
// Memory1: 9.36MB, beats 87.98%

// Runtime2: 0ms, beats 100%
// Memory2: 9.35MB, beats 87.98%

// Runtime3: 0ms, beats 100%
// Memory3: 9.29MB, beats 95.65%

File: Dynamic_Programming/198_House_Robber/solution198.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        rob1, rob2 = 0, 0

        for n in nums:
            temp = max(n + rob1, rob2)
            rob1 = rob2
            rob2 = temp
        return rob2

# Runtime1: 30ms, beats 89.63%
# Memory1: 16.44MB, beats 70.50%

# Runtime2: 34ms, beats 70.87%
# Memory2: 16.46MB, beats 70.50%

# Runtime3: 25ms, beats 97.97%
# Memory3: 16.56MB, beats 28.72%